### ISSUE TITLE
Notifications are sent once a day for each type

### ISSUE THREAD

**Issue by t-shehab** [zammad/zammad#4348]
### Infos:

* Used Zammad version: 5.2.3
* Installation method (source, package, ..): any
* Operating system: any
* Database + version: any
* Elasticsearch version: any
* Browser + version: any
* Ticket#: 10117576, Ticket#: 10120272, Ticket#10127540

### Expected behavior:

* When a ticket is set to "pending reminder" after it has already reached its first "pending reached" state, a notification is sent at 2nd "pending reached" state.

* Repeated notifications should be sent once a day to keep existing planned behavior.

### Actual behavior:

Escalation and pending reminder notifications are sent once a day. Even value triggering then notification has changed.

For example, after a ticket has reached the state "pending reached" for the first time, no further Email notification will be sent at subsequent "pending reached" states.

The original idea was to avoid too many repeated notifications. Currently notifications are re-sent every day in the early hours.

Related code:

https://github.com/zammad/zammad/blob/580495e9d95f05f77cad65d484c6f3c9fa25016c/app/models/transaction/notification.rb#L130

### Steps to reproduce the behavior:

* Make sure ticket owner's notification checkbox at "Also notify via email" for "Ticket reminder reached" is ticked.
* Set the ticket to "pending reminder".
* Wait for state "reminder reached" and for the notification Email being sent.
* Set another "pending reminder".
* No Email will be sent at the 2nd "reminder reached" state.

Note: the above behaviour will also occur if you set the ticket to another state first (like: "open") before setting a 2nd "pending reminder".

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by t-shehab** [comment]
An online notification also does not get sent when a reminder is set after first reminder was reached.

This is in 5.3.

Ticket#10120272

**Comment by barfuss284** [comment]
Hi, same problem here with 5.4 - how to fix this?

**Comment by MrGeneration** [comment]
Please be patient for this issue to be fixed. You'll see when it's closed and going to stable.

**Comment by mantas** [comment]
@t-shehab do you mean not sending pending reminder again in the same day? There's a limit of one-pending-reminder-per-day. Pending reminders will be repeated every day till pending reminder is solved. This does seem to work fine.

**Comment by t-shehab** [comment]
@mantas yes, that's exactly what I mean. For some use-cases it is important to have the possibility of multiple reminders per day. I think this is a bug, because nowhere does the software tell me that a subsequent reminder reached on the same day will not trigger another notification.

As a user, if I get a notification for reminder reached and decide to then set the pending reminder to a later time on the same day, I expect to be notified at that time, as well. Otherwise, the software should at least tell me upfront that it won't remind me more than once per day. That would still be bad, imo, but not as bad as just quietly not notifying me.

**Comment by mantas** [comment]
This is the related code bit - https://github.com/zammad/zammad/blob/580495e9d95f05f77cad65d484c6f3c9fa25016c/app/models/transaction/notification.rb#L130

IMO this should go back to `needs specification` for a holistic solution. I remember there were good reasons why we didn't touch this the previous time we bumped into this.

@dominikklein @mgruner any opinions?

**Comment by mgruner** [comment]
I also think it needs to go through specification. It's been like this probably from the start of Zammad.

**Comment by mantas** [comment]
I edited the original issue to reflect the under-the-hood problem better and set it back to `specification requited`

**Comment by RepinsPL** [comment]
Problem went bigger after implementing: https://github.com/zammad/zammad/issues/4484
Now it's problem not only for notifications, but for whole code execution of triggers which may contain business critical logic.

Proposed solution:
After line:
https://github.com/zammad/zammad/blob/580495e9d95f05f77cad65d484c6f3c9fa25016c/app/models/transaction/notification.rb#L130

Add condition, that if type == 'reminder_reached' then already_notified_cutoff is a maximum of current value and last modification time of 'pending till' value (also read from history).

This one-liner simple change will allow workflows like:
1) Agent set "pending till" for "11:00 am"
2) When reminder is reached Agent wants to postpone ticket for 1 hour later and sets "pending till" for "12:00 am"
3) When second reminder is reached notification after proposed change will be created despite fact, that it is the same day
